### ISSUE TITLE
Is `AWS_PROFILE` being cached?

### ISSUE THREAD

**Issue by hutch3232** [fsspec/s3fs#806]
Hello, I'm new to using this package (and python) and I do see some similar issues (linked at the bottom), but I think this demonstrates some unusual behavior. I hope this is helpful.

FWIW I'm using an S3 Compatible Storage called [Scality](https://www.scality.com/ring/).

**Goal:**
I want to be able to set the `AWS_PROFILE` environmental variable to a particular profile that has a record in my `~/.aws/credentials` and `~/.aws/config` file. Then I would like `s3fs` to pick up on that variable and use it for subsequent commands. Then if I change the profile (by overriding the `AWS_PROFILE` environmental variable), I'd like subsequent `s3fs` calls to use that one. 

**Issue:**
The first half of my goal works. I can set the `AWS_PROFILE` to a non-default profile and `s3fs` will pick it up correctly. However, if I change the environmental variable, `s3fs` continues to try to use the original profile.

**Note:**
I'm including `pandas` S3 reads in here to demonstrate how, while the `cache_regions` fix (mentioned in the linked issues) does help `s3fs` its use within `pandas` doesn't seem to be getting picked up.

```python
import os
import s3fs
import pandas as pd

# IF you have a default profile, this works, otherwise does not work (expected)
# i.e., if you have a default profile it will use it, otherwise it doesn't know so it will error, which is good
s3 = s3fs.S3FileSystem()
s3.ls("first-bucket")
pd.read_csv("s3://first-bucket/file.csv")
# <restart kernel before proceeding>

# works
# here we can see my first attempt at setting the env var is successful for s3fs and pandas
os.environ["AWS_PROFILE"] = "first-bucket-profile"
s3 = s3fs.S3FileSystem()
s3.ls("first-bucket")
pd.read_csv("s3://first-bucket/file.csv")

# does not work
# now when I go to change the env var, s3fs doesn't respect it (pandas doesn't either)
s3.ls("second-bucket") # expected to not work and indeed it does not
os.environ["AWS_PROFILE"] = "second-bucket-profile"
s3 = s3fs.S3FileSystem()
s3.ls("second-bucket")
pd.read_csv("s3://second-bucket/example.csv")
# ClientError: An error occurred (403) when calling the HeadObject operation: Forbidden

# works
# within the same kernel, if I update the profile env var AND say cache_regions=False
# it works for s3fs but not for pandas
os.environ["AWS_PROFILE"] = "second-bucket-profile"
s3 = s3fs.S3FileSystem(cache_regions = False)
s3.ls("second-bucket")
pd.read_csv("s3://second-bucket/example.csv") # does not work (same ClientError)

# works
# alternatively I did find that I could just set profile and of course it works but pandas does not
# unless I give it a profile in storage_options
s3 = s3fs.S3FileSystem(profile = "second-bucket-profile")
s3.ls("second-bucket")
pd.read_csv("s3://second-bucket/example.csv") # does not work (same ClientError)
pd.read_csv("s3://second-bucket/example.csv", storage_options={"profile": "second-bucket-profile"})
```

In summary, the first case where I set the env var works perfectly. I just wish it would work to simply change that env var a second time so that I can easily switch my `pandas` reads between profiles.

#701 #754

**Comment by martindurant** [comment]
s3fs does not store the value of the environment variable, but it *does* store the S3FileSystem instances themselves. Thus, if you do S3FileSystem(...) with the same arguments as a previous instance, you will get that exact same instance back instead of creating a new one (this saves on various connection overheads). 

To prevent the behaviour, you can pass `skip_instance_cache=True` for each instantiation, or set `s3fs.S3FileSystem.cachable = False` for the duration of the session.

https://filesystem-spec.readthedocs.io/en/latest/features.html#instance-caching

**Comment by hutch3232** [comment]
This worked great for me, thanks for the awesome package and quick response!
